src/fenics_adjoint/types/function.py

```diff
@@ -208,6 +208,35 @@ def _ad_convert_type(self, value, options=None):
             raise NotImplementedError(
                 "Unknown Riesz representation %s" % riesz_representation)
 
+    def _define_riesz_map_form(self, riesz_map):
+        u = dolfin.TrialFunction(self.function_space())
+        v = dolfin.TestFunction(self.function_space())
+        if riesz_map == "L2":
+            a = dolfin.inner(u, v) * dolfin.dx
+        elif riesz_map == "H1":
+            a = dolfin.inner(u, v)*dolfin.dx \
+                + dolfin.inner(dolfin.grad(u), dolfin.grad(v))*dolfin.dx
+        else:
+            raise NotImplementedError(
+                "Unknown Riesz representation %s" % riesz_map)
+        return a
+    
+    def _ad_riesz_representation(self, options=None):
+        options = {} if options is None else options
+        riesz_map = options.get("riesz_map", "l2")
+
+        if riesz_map == "l2":
+            return self
+        elif riesz_map in ("L2", "H1"):
+            a = self._define_riesz_map_form(riesz_map)
+            covector = dolfin.assemble(dolfin.action(a, self))
```

Comment:
I agree that an `enum` is a better way. In the latest commit, I added the `NormType`:

```
class NormType(str, Enum):
    """Norm types for Riesz representation.
    """
    l2 = 'l2'
    L2 = 'L2'
    H1 = 'H1'
```

I also considered the following implementation:

```
class NormType( Enum):
    """Norm types for Riesz representation.
    """
    l2 = 1
    L2 = 2
    H1 = 3
```

However, this would need the `NormType` directly in the front end and we would modify for instance this [example](https://github.com/dolfin-adjoint/dolfin-adjoint/blob/main/examples/stokes-topology/stokes-topology-rol.py) at line [#L241](https://github.com/dolfin-adjoint/dolfin-adjoint/blob/main/examples/stokes-topology/stokes-topology-rol.py#L241) to the form:

```python
solver = ROLSolver(problem, params, inner_product=NormType.L2)
```

Thus, I opted to keep `NormType(str, Enum)`, eliminating the need for frontend adjustments. Does that sound good to you?

Reply:
That works for me.